### PR TITLE
Fix FP magic in Tinflex algorithm

### PR DESCRIPTION
**Pull request by wilzbach**
@9il @joseph-wakeling-sociomantic I had to wade through quite a lot of issues to get Tinflex working on 32-bit. With this PR all Tinflex tests should pass :)

I also fixed the logging and I would prefer if we can keep the hex dump helpers, they are highly useful.

### CODE REVIEW

**Comment by 9il** [comment]
Thanks, will review it tomorrow

**Comment by wilzbach** [comment]
>  I had to wade through quite a lot of issues to get Tinflex working on 32-bit. With this PR all Tinflex tests should pass :)

The good news
- all 15 jobs on Travis are green (64 & 32-bit of DMD and LDC on Linux)
- Windows 64-bit is passing too

Bad news:
- Windows 32-bit + Windows LDC are failing

File: .travis.yml [hunk 1]
```diff
@@ -18,11 +18,11 @@ matrix:
     - {os: linux, d: dmd-2.069.2, env: ARCH="x86", addons: {apt: {packages: [[gcc-multilib]]}}}
     - {os: linux, d: dmd-2.068.2, env: ARCH="x86", addons: {apt: {packages: [[gcc-multilib]]}}}
     - {os: linux, d: ldc-1.0.0, env: ARCH="x86", addons: {apt: {packages: [[gcc-multilib]]}}}
-    - {os: linux, d: ldc-0.17.0, env: ARCH="x86", addons: {apt: {packages: [[gcc-multilib]]}}}
 branches:
   only:
     - master
     - stable
+    - tinflex
```
Comment (wilzbach) [inline .travis.yml:21]: modification are only done temporarily to enable Travis & Co for the `tinflex` branch.

They will be removed before `tinflex` gets merged into master.

File: source/mir/random/flex/internal/transformations.d [hunk 2]
```diff
@@ -54,7 +62,8 @@ in
 }
 body
 {
-    import mir.internal.math: pow, exp, copysign;
+    import mir.internal.math: pow, copysign;
+    import std.math : exp;
```
Comment (wilzbach) [inline source/mir/random/flex/internal/transformations.d]: `exp` and llvm `exp` yield different results.

Comment (9il) [inline source/mir/random/flex/internal/transformations.d]: Does it crash the tests?

File: source/mir/random/flex/internal/area.d [hunk 3]
```diff
@@ -193,12 +202,18 @@ body
     {
         if (fabs(z) < constants!S.smallExp)
         {
-            area = exp(sh.a) * (ivLength) * (1 + z * S(0.5) + (z * z) * one_div_6
-                                               + (z * z * z) * one_div_24);
+            import std.math : std_exp = exp;
```
Comment (9il) [inline source/mir/random/flex/internal/area.d]: ditto

File: source/mir/random/flex/internal/transformations.d (hunk 2 above)
Comment (wilzbach) [inline source/mir/random/flex/internal/transformations.d]: that's how I know about it. Only a tiny difference can yield a different splitting in the Tinflex algorithm :/

File: source/mir/random/flex/internal/area.d (hunk 3 above)
Comment (wilzbach) [inline source/mir/random/flex/internal/area.d]: I will try once again, now that all tests are passing, but last time it did make the difference between passing and failing.

File: source/mir/random/flex/internal/transformations.d [hunk 4]
```diff
@@ -203,7 +244,13 @@ S antiderivative(bool common = false, S)(in S x, in S c)
             return -log(-x);
     }
     auto d = c + 1;
-    return fabs(c) / d * pow(fabs(x), d / c);
+    // std.math.pow and llvm_pow are different
+    import std.math : pow;
+    // surpress DMD's annoying FP magic
```
Comment (9il) [inline source/mir/random/flex/internal/transformations.d:246]: What test are failing with LLVM pow?

File: source/mir/random/flex/internal/area.d (hunk 3 above)
Comment (wilzbach) [inline source/mir/random/flex/internal/area.d]: if I revert back to `powf` (the LDC version), I get

```
2016-08-01T18:13:26.432:area.d:__foreachbody9:565 hat: 0xc.9bbfbb4104cf35p+2, squeeze: 0xa.2412372f691d7b9p-5
2016-08-01T18:13:26.432:area.d:__foreachbody9:566 exp. hat: 0xc.9bbfbb4104cf353p+2, squeeze: 0xa.2412372f691d7bbp-5
2016-08-01T18:13:26.432:area.d:__foreachbody9:567 iv: Interval!real(-0xcp-2, -0xcp-3, 0x0p+0, -0xap+2, 0x9.cp+3, -0xc.4p+3, 0x8.cp-2, -0xcp-3, -0x8.8p+1, LinearFun!real(-0xcp-3, -0xcp-3, 0x8.cp-2), LinearFun!real(0xe.1p+1, -0xcp-3, 0x8.cp-2), 0xc.9bbfbb4104cf35p+2, 0xa.2412372f691d7b9p-5)
core.exception.AssertError@source/mir/random/flex/internal/area.d(586): Assertion failure
----------------
```

File: source/mir/random/flex/internal/transformations.d (hunk 2 above)
Comment (9il) [inline source/mir/random/flex/internal/transformations.d]: Yes, and tests should be OK about that.

Comment (wilzbach) [inline source/mir/random/flex/internal/transformations.d]:
```
2016-08-01T18:15:41.235:area.d:__foreachbody9:467 hat: 0x1.d3c24e4b0f649p+4, squeeze: 0x1.d635b6e68a728p-58
2016-08-01T18:15:41.235:area.d:__foreachbody9:468 exp. hat: 0x1.d3c24e4b0f649p+4, squeeze: 0x1.d635b6e68a727p-58
2016-08-01T18:15:41.235:area.d:__foreachbody9:469 iv: Interval!double(-0x1.8p+1, -0x1.8p+0, 0x1.ccccccccccccdp-1, 0x1.0b6c3afdde064p-52, 0x1.255450b4dc554p-46, 0x1.3bfcb71aca3c1p-40, 0x1.ca5999918a498p+2, -0x1.3562e1423d581p+3, -0x1.8216980998491p+6, LinearFun!double(-0x1.3562e1423d581p+3, -0x1.8p+0, 0x1.ca5999918a498p+2), LinearFun!double(0x1.255450b4dc554p-46, -0x1.8p+1, 0x1.0b6c3afdde064p-52), 0x1.d3c24e4b0f649p+4, 0x1.d635b6e68a728p-58)
core.exception.AssertError@source/mir/random/flex/internal/area.d(479): Assertion failure
----------------
```

File: source/mir/random/flex/internal/area.d (hunk 3 above)
Comment (9il) [inline source/mir/random/flex/internal/area.d]: Is it exact `==` or approxEqual?

File: source/mir/random/flex/internal/transformations.d (hunk 4 above)
Comment (wilzbach) [inline source/mir/random/flex/internal/transformations.d:246]:
```
2016-08-01T18:17:52.276:area.d:__foreachbody9:567 hat: 0xf.fffffffffffffffp+16380, squeeze: 0xb.0542494690fac2ap-59
2016-08-01T18:17:52.276:area.d:__foreachbody9:568 exp. hat: 0xf.fffffffffffffffp+16380, squeeze: 0xb.0542494690fac27p-59
2016-08-01T18:17:52.276:area.d:__foreachbody9:569 iv: Interval!real(-0xcp-2, -0xcp-3, -0xcp-3, -0xb.cede4ee29e33249p+83, 0xa.cb17341b249abf7p+90, -0x9.f8c25c3dfff8d07p+97, -0x9.9eecc36f0775795p-8, -0xa.d2ca5bdce864288p-7, -0x9.304920a78a9fb26p-3, LinearFun!real(-0xa.d2ca5bdce864288p-7, -0xcp-3, -0x9.9eecc36f0775795p-8), LinearFun!real(0xf.be7dbe837d99861p+82, -0xcp-3, -0x9.9eecc36f0775795p-8), 0xf.fffffffffffffffp+16380, 0xb.0542494690fac2ap-59)
core.exception.AssertError@source/mir/random/flex/internal/area.d(588): Assertion failure
----------------
```

File: source/mir/random/flex/internal/area.d (hunk 3 above)
Comment (wilzbach) [inline source/mir/random/flex/internal/area.d]: Exact `==`

File: source/mir/random/flex/internal/transformations.d (hunk 4 above)
Comment (9il) [inline source/mir/random/flex/internal/transformations.d:246]: ditto, `==` or approxEqual?

File: source/mir/random/flex/internal/area.d (hunk 3 above)
Comment (9il) [inline source/mir/random/flex/internal/area.d]: change to approxEqual (for all tests)

File: source/mir/random/flex/internal/transformations.d (hunk 2 above)
Comment (wilzbach) [inline source/mir/random/flex/internal/transformations.d]: > Yes, and tests should be OK about that.

So you want to use `approxEqual` everywhere?

File: source/mir/random/flex/internal/area.d (hunk 3 above)
Comment (9il) [inline source/mir/random/flex/internal/area.d]: BTW, how you obtained values for this test?

File: source/mir/random/flex/internal/area.d [hunk 5]
```diff
@@ -3,6 +3,53 @@ module mir.random.flex.internal.area;
...
+        {
+            version = Flex_fpEqual;
+        }
+    }
+    version(Flex_fpEqual)
+    {
+        import std.math : approxEqual;
+        bool fpEqual(float a, float b) { return a.approxEqual(b, 1e-8, 1e-8); }
+        bool fpEqual(double a, double b) { return a.approxEqual(b, 1e-16, 1e-16); }
+        bool fpEqual(real a, real b) { return a.approxEqual(b, 1e-18, 1e-18); }
```
Comment (wilzbach) [inline source/mir/random/flex/internal/area.d:59]: why is something like this not the default? :/

**Comment by wilzbach** [comment]
FYI tests are passing on Win32 too :tada: 
(it was just a weird issue with logf)

FP is still crazy for LDC on Windows, but this a known issue :/

File: source/mir/random/flex/internal/area.d [hunk 6]
```diff
@@ -764,3 +1037,68 @@ unittest
         }
     }
 }
+
+/+
+// used to dump data tumples
```
Comment (wilzbach) [inline source/mir/random/flex/internal/area.d:1042]: this will be removed in the last cleanup step of Tinflex

**Pull request merged**